### PR TITLE
fix(escapeAllSwigTags): check tag completeness

### LINKED ISSUE #5301: Incomplete display of HTML document, truncated at '#' in markdown document
使用默认安装，将md文档处理为网页时，启动hexo网站服务，本地查看文档内容不全。
<br>When using the default installation to process MD documents as web pages, starting the hexo website service may result in incomplete local viewing of document content.

``` md
### 数组

- 数组定义：`array_name=(value1 ... valuen)`<br>
  括号里面允许换行，value以空格或换行分隔
- 也可以单独定义数组各个分量：`array_name[i]=valuei` (i可以是数字).<br>
  即数组可以不定义就使用，也可以定义时赋初值，数组元素就像普通变量一样使用即可。
- 引用数组所有元素：`${array_name[*]}`或`${array_name[@]}`.<br>
  只会取得数组已定义元素值的列表
- 可以使用unset删除数组元素：`unset array_name[i]`<br>
- 使用`${#array_name[*]}`或`${#array_name[@]}`取得数组已定义的元素数量(数组长度)
- 使用`${#array_name[n]}`取得n号元素的长度，这些就像求串长一样。
- 数组切片：`${array_name[*]:position:length}`或`${array_name[@]:position:length}`.<br>
  这其实是对数组已定义元素值的列表进行切片，就像串取子串一样。
```

上面文档将只会显示部分处理内容:
<br>The above document will only display partial processing content:

``` md
### 数组

- 数组定义：`array_name=(value1 ... valuen)`<br>
  括号里面允许换行，value以空格或换行分隔
- 也可以单独定义数组各个分量：`array_name[i]=valuei` (i可以是数字).<br>
  即数组可以不定义就使用，也可以定义时赋初值，数组元素就像普通变量一样使用即可。
- 引用数组所有元素：`${array_name[*]}`或`${array_name[@]}`.<br>
  只会取得数组已定义元素值的列表
- 可以使用unset删除数组元素：`unset array_name[i]`<br>
- 使用`$
```

经过多次验证，只会在正文包含'#'字符时出现异常。
<br>After multiple verifications, exceptions will only occur when the body contains the '#' character.

版本信息：
```
npx hexo -v
INFO  Validating config
hexo: 6.3.0
hexo-cli: 4.3.1
os: win32 10.0.19044
node: 18.17.1
acorn: 8.8.2
ada: 2.5.0
ares: 1.19.1
brotli: 1.0.9
cldr: 43.1
icu: 73.2
llhttp: 6.0.11
modules: 108
napi: 9
nghttp2: 1.52.0
openssl: 3.1.2
simdutf: 3.2.12
tz: 2023c
undici: 5.22.1
unicode: 15.0
uv: 1.46.0
uvwasi: 0.0.18
v8: 10.2.154.26-node.26
zlib: 1.3
```

### PR DESCRIPTION
**Pull request by uiolee**
## What does it do?

fix #5301
fix https://github.com/hexojs/hexo-renderer-pandoc/issues/60

Before escape tags, check if it contains a complete tag.
In others words, just escape complete tags.

## Pull request tasks

- [x] Add test cases for the changes.
- [x] Passed the CI test.

### CODE REVIEW

File: lib/hexo/post.ts [hunk 1]
```diff
@@ -84,13 +84,14 @@ class PostRenderEscape {
 
       if (state === STATE_PLAINTEXT) { // From plain text to swig
         if (char === '{') {
-          if (next_char === '{') {
+          // check if it is a complete tag {{ }}
+          if (next_char === '{' && /\{\{.+?\}\}/.test(str)) {
             state = STATE_SWIG_VAR;
             idx++;
-          } else if (next_char === '#') {
+          } else if (next_char === '#' && /\{#.+?#\}/.test(str)) {
```
Comment (SukkaW) [inline lib/hexo/post.ts:91]: With the change, we will run `regexp.test` again and again for every character inside `str`.

The `.test` should be hoist to the top of the function.

**Review by SukkaW — CHANGES_REQUESTED** [review]

File: lib/hexo/post.ts (hunk 1 above)
Comment (uiolee) [reply lib/hexo/post.ts]: thanks, I have updated it

**Review by SukkaW — APPROVED** [review]

**Comment by appotry** [comment]
7.1 没有完全修复，解析下面代码会报错。7.0 反而解析正常。

Version 7.1 is not fully fixed; parsing the following code will result in an error. Surprisingly, parsing is normal in version 7.0

```
{% valkyrurl
[url=/posts/f4926c88/]
[title="RSS的使用与Tiny Tiny RSS Selfhost自建"]
[avatar=/medias_webp/cover/rss.webp]
[desc="RSS是一种消息来源的格式规范，网站可以按照这种格式规范提供文章的标题、摘要、全文等信息给订阅用户，用户可以通过订阅不同网站 RSS 链接的方式将不同的信息源进行聚合，在一个工具里阅读这些内容"]
%}
```

err log

```
13:09:23.020 FATAL Something's wrong. Maybe you can find the solution here: https://hexo.io/docs/troubleshooting.html
Template render error: (unknown path)
  TypeError: Cannot read properties of null (reading '1')
    at Object._prettifyError (/app/node_modules/nunjucks/src/lib.js:32:11)
    at /app/node_modules/nunjucks/src/environment.js:464:19
    at eval (eval at _compile (/app/node_modules/nunjucks/src/environment.js:527:18), <anonymous>:11:11)
    at tryCatcher (/app/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/app/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/app/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromiseCtx (/app/node_modules/bluebird/js/release/promise.js:641:10)
    at _drainQueueStep (/app/node_modules/bluebird/js/release/async.js:97:12)
    at _drainQueue (/app/node_modules/bluebird/js/release/async.js:86:9)
    at Async._drainQueues (/app/node_modules/bluebird/js/release/async.js:102:5)
    at Async.drainQueues (/app/node_modules/bluebird/js/release/async.js:15:14)
    at process.processImmediate (node:internal/timers:478:21)
```

**Comment by appotry** [comment]
7.1版本不止上面这一个tag 报错，7.0反而正常

**Comment by D-Sketon** [comment]
> 7.1 没有完全修复，解析下面代码会报错。7.0 反而解析正常。
> 
> Version 7.1 is not fully fixed; parsing the following code will result in an error. Surprisingly, parsing is normal in version 7.0
> 
> ```
> {% valkyrurl
> [url=/posts/f4926c88/]
> [title="RSS的使用与Tiny Tiny RSS Selfhost自建"]
> [avatar=/medias_webp/cover/rss.webp]
> [desc="RSS是一种消息来源的格式规范，网站可以按照这种格式规范提供文章的标题、摘要、全文等信息给订阅用户，用户可以通过订阅不同网站 RSS 链接的方式将不同的信息源进行聚合，在一个工具里阅读这些内容"]
> %}
> ```
> 
> err log
> 
> ```
> 13:09:23.020 FATAL Something's wrong. Maybe you can find the solution here: https://hexo.io/docs/troubleshooting.html
> Template render error: (unknown path)
>   TypeError: Cannot read properties of null (reading '1')
>     at Object._prettifyError (/app/node_modules/nunjucks/src/lib.js:32:11)
>     at /app/node_modules/nunjucks/src/environment.js:464:19
>     at eval (eval at _compile (/app/node_modules/nunjucks/src/environment.js:527:18), <anonymous>:11:11)
>     at tryCatcher (/app/node_modules/bluebird/js/release/util.js:16:23)
>     at Promise._settlePromiseFromHandler (/app/node_modules/bluebird/js/release/promise.js:547:31)
>     at Promise._settlePromise (/app/node_modules/bluebird/js/release/promise.js:604:18)
>     at Promise._settlePromiseCtx (/app/node_modules/bluebird/js/release/promise.js:641:10)
>     at _drainQueueStep (/app/node_modules/bluebird/js/release/async.js:97:12)
>     at _drainQueue (/app/node_modules/bluebird/js/release/async.js:86:9)
>     at Async._drainQueues (/app/node_modules/bluebird/js/release/async.js:102:5)
>     at Async.drainQueues (/app/node_modules/bluebird/js/release/async.js:15:14)
>     at process.processImmediate (node:internal/timers:478:21)
> ```
> 

见＃5401

**Comment by appotry** [comment]
> > 7.1 没有完全修复，解析下面代码会报错。7.0 反而解析正常。
> > Version 7.1 is not fully fixed; parsing the following code will result in an error. Surprisingly, parsing is normal in version 7.0
> > ```
> > {% valkyrurl
> > [url=/posts/f4926c88/]
> > [title="RSS的使用与Tiny Tiny RSS Selfhost自建"]
> > [avatar=/medias_webp/cover/rss.webp]
> > [desc="RSS是一种消息来源的格式规范，网站可以按照这种格式规范提供文章的标题、摘要、全文等信息给订阅用户，用户可以通过订阅不同网站 RSS 链接的方式将不同的信息源进行聚合，在一个工具里阅读这些内容"]
> > %}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > err log
> > ```
> > 13:09:23.020 FATAL Something's wrong. Maybe you can find the solution here: https://hexo.io/docs/troubleshooting.html
> > Template render error: (unknown path)
> >   TypeError: Cannot read properties of null (reading '1')
> >     at Object._prettifyError (/app/node_modules/nunjucks/src/lib.js:32:11)
> >     at /app/node_modules/nunjucks/src/environment.js:464:19
> >     at eval (eval at _compile (/app/node_modules/nunjucks/src/environment.js:527:18), <anonymous>:11:11)
> >     at tryCatcher (/app/node_modules/bluebird/js/release/util.js:16:23)
> >     at Promise._settlePromiseFromHandler (/app/node_modules/bluebird/js/release/promise.js:547:31)
> >     at Promise._settlePromise (/app/node_modules/bluebird/js/release/promise.js:604:18)
> >     at Promise._settlePromiseCtx (/app/node_modules/bluebird/js/release/promise.js:641:10)
> >     at _drainQueueStep (/app/node_modules/bluebird/js/release/async.js:97:12)
> >     at _drainQueue (/app/node_modules/bluebird/js/release/async.js:86:9)
> >     at Async._drainQueues (/app/node_modules/bluebird/js/release/async.js:102:5)
> >     at Async.drainQueues (/app/node_modules/bluebird/js/release/async.js:15:14)
> >     at process.processImmediate (node:internal/timers:478:21)
> > ```
> 
> 见 #5401，和换行有关

那个issue看过，找到这里来的

文章太多，不可能一个个tag修改过去， 只能先降级到7.0版本使用了。等待后面修复

**Pull request merged**